feat(docs): Document slow update tree (#3416)
AztecProtocol/aztec-packages: docs/docs/concepts/foundation/communication/public_private_calls/slow_updates_tree.md

---
title: Privately access Historical Public Data
---

> In Aztec, private and public execution environments are completely separate and operate with distinct state management. It is not possible for private functions to reliably access the most recent public data public state - only sequencers can do that.

Reviewer: can you add a prerequisite to read the previous document first (public->private calls)

- Access Control: Managing privileges in contracts, such as a token contract owner’s ability to mint new tokens, is streamlined when control information is shared between public and private executions. This might include things like blacklists and whitelists.

## How it works

We developed the Slow Updates Tree to help balance public and private execution in a blockchain context. Earlier systems typically used either fully public or entirely private state trees.

**First iteration: Shared State Tree** 

Using a shared state tree, it is possible to privately access historical public data by providing a membership proof to show that a value is indeed part of a commitment, and then check that the commitment matches the one stored in the state, as shown below.

```mermaid
graph TD;
    Cm[Commmitment1] --> Vm1[Value1]
    Cm --> Vm2[Value2]
    Cm --> Vmn[Value3]
    Vm1 --> Mp1[Membership Proof for Value1]
    Mp1 --> St[Commitment in State]
    St --> Cm
```

Note: a *commitment* refers to a cryptographic assurance that a specific set of data is included in the tree. In the current design, we are using Merkle trees to commit to the values.

> However, these would be contract-specific trees, so we will be leaking the contract address of the contract doing a lookup. 

Author: should i do single tree -> slow updates tree? & just skip the middle one

> **Second iteration: Multi-level Shared State Tree**

Reviewer: This is the 2nd iteration but rather the final one and doesn't exist yet. Best to delete.

This privacy issue can be solved by organizing data into layers, with each contract having its own commitment within a larger tree. Only the top-level commitment is revealed.

```mermaid
graph TD;
    C[Top Commitment] -->|Contains| C1[Commitment1]
    C -->|Contains| Cm[Commitment2]
    Cm -->|Contains| Vm1[Value1]
    Cm -->|Contains| Vm2[Value2]
    Cm -->|Contains| Vmn[Valuen]
```

However, this means that any canges in any contract's data will change the top-level commitment and invalidate the proofs for all other contracts.

**Third iteration: Slow Updates Tree**

> To solve this, the Slow Updates Tree is a dual-tree structure - a current tree and a pending tree. Any updates are added to the pending tree, which then becomes the current tree at the end of an epoch. The pending tree is replicated from the current tree, and the cycle begins again.

Author: lasse's doc seems to suggest it does solve this problem - maybe i misunderstood? or is hasnt been implemented that way?

---
title: Privately access Public State
---

> In Aztec, private and public execution environments are completely separate and operate with distinct state management. In the previous section we learned that functions can call private functions, and public functions can save messages to a merkle tree to then be read by a private function.

Reviewer: > that functions can call private functions
?

> In Aztec, private and public execution environments are completely separate and operate with distinct state management. In the previous section we learned that functions can call private functions, and public functions can save messages to a merkle tree to then be read by a private function.

Reviewer: ```suggestion
In Aztec, private and public execution environments are completely separate and operate with distinct state management. It is not possible for private functions to reliably access the most recent public data public state (only sequencers can do that!)
```

> It is also possible to access persistent data across both private and public state using a **slow updates tree**, which we will talk about on this page. Please note that we are still experimenting with this feature and it is not something that has ever been achieved before.

Reviewer: Uhh so slow updates doesn't let you read public state due to race conditions. And we **need** to be very clear to prevent people from thinking this:

```suggestion
But, what about historical public data? Through a **slow updates tree**, you can have private functions access historical public state. Please note that we are still experimenting with this feature.
```

On this page you will learn:

1. Why a slow updates tree exists & use cases
2. How it works

> 3. How it can be used to access data in public & private domains

Reviewer: ```suggestion
3. How it can be used to access historical public data in private 
```

4. Limitations

> ## The need for a slow updates tree

Reviewer: really like the idea behind this section! But need to fix the "access data across private and public" connotation

Author: fixed

---
title: Privately access Historic Public Data
---

In Aztec, private and public execution environments are completely separate and operate with distinct state management. It is not possible for private functions to reliably access the most recent public data public state - only sequencers can do that.

But, what about historical public data (or public data that changes infrequently)? Through a **slow updates tree**, private functions can access historical public state. Please note that we are still experimenting with this feature.

On this page you will learn:

1. Why a slow updates tree exists & use cases
2. How it works
3. How it can be used to access historical public data
4. Limitations

## The need for a slow updates tree

> This structure was created specifically to privately & publicly access historical public data that is not updated often. It should be used to store data that is not sensitive but needs to be accessible from private smart contracts. Using a slow updates tree, it is possible to:

Author: lmao i shall reword

- Access historic public data from a private function
- Access historic public data from a public function
- Update public data (that does not need updated often) from public and private functions 

This data structure is ideal for these use cases:

- Address Registry: Enabling contracts to interact with other contracts more easily requires address storage accessible in both public and private executions. This can be particularly helpful in things such as proxy contracts.
- Access Control: Managing privileges in contracts, such as a token contract owner’s ability to mint new tokens, is streamlined when control information is shared between public and private executions. This might include things like blacklists and whitelists.

## How it works

We developed the Slow Update Tree to help balance public and private execution in a blockchain context.

Earlier systems typically used either fully public or entirely private state trees, which led to privacy leaks if reading public data from private state.

> Using a shared state tree, it is possible to privately access historic public data by providing a membership proof to show that a value is indeed part of a commitment, and then check that the commitment matches the one stored in the state, as shown below.

Author: im kinda using commitment to mean root. maybe i need to explain this better.

 a merkle tree is the current design but since we're talking about shared trees in general it doesnt have to be. to quote lasse - Generally, the design relies on vector commitments where the values and commitment are publicly available in such a manner that they can be easily read in public and used for inclusion checks in private. Logically, you can think of a contract specific SST as what is shown in Figure 1. Where we have a commitment Cm to a set of values Vm,j ∀j for a specific contract m. In the current design we are using Merkle Trees to commit the values.... We can provide a membership proof for Vm,j ∈Cm to show that the value is indeed part of the commitment, and then check that the commitment matches the one stored in the state

```mermaid
graph TD;
    Cm[CommitmentToKnowingAValue] --> Vm1[Value1]
    Cm --> Vm2[Value2]
    Cm --> Vmn[Value3]
    Vm1 --> Mp1[Membership Proof for Value1]
    Mp1 --> St[Commitment in State]
    St --> Cm
```

However this means that any changes to the commitment will invalidate our membership proof, and make our read fail.

> To solve this, the Slow Update Tree is a dual-tree structure - using a current and pending tree - that updates at the end of an epoch. This way, we can ensure that the values are stable throughout the epoch, and that the membership proofs are not invalidated by changes in other contracts more than once every epoch.

Reviewer: to prevent confusion down the line, can you explain the use of the pending tree -> that any updates that need to be made go into the pending tree and after an epoch, they become the current tree. Anf then say, by having two separate trees, we can ensure the valyes are stable...

```mermaid
graph TD;
    Change{Epoch Over} -->|True| Current{Current}
    Change -->|False| Pending{Pending}
    Current --> Current1[Current Commitment 1]
    Current --> CurrentM[Current Commitment 2]
    CurrentM --> Value1[Current Value 1]
    CurrentM --> Value2[Current Value 2]
    CurrentM --> ValueN[Current Value n]
    Pending --> PendingM[Pending Commitment 1]
    PendingM --> PValue1[Pending Value 1]
    PendingM --> PValue2[Pending Value 2]
    PendingM --> PValueN[Pending Value n]
```

> ### Dual Tree Structure

Reviewer: redundant topic can be deleted

The Slow Update Tree operates on a dual-tree mechanism - a _current_ tree and a _pending_ tree. These trees manage the state data that needs to be shared between public and private executions.

This _slow update_ mechanism minimizes the invalidation of reads and writes due to concurrent updates.

### Epoch-Based Updates

> The current tree is replaced with the pending tree at the end of each epoch. Then a new pending tree is created, the epoch restarts, and the cycle begins again.

Reviewer: ```suggestion
The current tree is replaced with the pending tree at the end of each epoch. Then a new pending tree is replicated, the epoch restarts, and the cycle begins again.
```

### Accessing Data

**From public state:** Accessed directly from the state

> **From private state:** Performs a membership proof for the values in the tree, ensuring that they are part of the commitment.

Reviewer: again why commitment? In Aztec, commitment was so confusing for everyone that we moved away to notes and note hashes. Let's not reintroduce it. It is simply a tree

> 

Reviewer: add link to where people can see an example of accessing data

### Updating Values

> Updates are made to the pending tree. Then at the end of each epoch, the updates in the pending tree are committed and it becomes the current tree.

Reviewer: add link to where people can see an example of updating  data

## Limitations

### Delayed State Finality

Updates in the Slow Update Tree are only finalized at the end of an epoch.

> This could potentially lead to delays in state changes - for example an address may be added to a blacklist but this will not be reflected in the state until the end of the current epoch.

Reviewer: I wouldn't call this a limitation but more like a feature. The tree is only useful for historic data for a reason

### Complexity in State Management

> Developers are used to instant state updates, so the Slow Update Tree might take some getting used to. But we believe this won't take long!

Reviewer: this should be in the previous heading.

I would think "Complexity in State Management" is more about how you need to manage trees
